File: Problems/Algorithms/567. Permutation in String/PermutationString.kt
class Solution {
    fun checkInclusion(s1: String, s2: String): Boolean {
        val n1 = s1.length
        val n2 = s2.length
        if (n1 > n2) return false
        
        val map1 = IntArray(26)
        val map2 = IntArray(26)
        
        for (i in 0..n1-1) {
            map1[s1.get(i)-'a']++
            map2[s2.get(i)-'a']++
        }
        
        for (i in 0..n2-n1-1) {
            if (match(map1, map2)) return true
            
            map2[s2.get(i+n1)-'a']++
            map2[s2.get(i)-'a']--
        }
        
        return match(map1, map2)
    }
    
    private fun match(map1: IntArray, map2: IntArray): Boolean {
        for (i in 0..25) {
            if (map1[i] != map2[i]) {
                return false
            }
        }
        
        return true
    }
}

File: Problems/Algorithms/567. Permutation in String/permutation_string.rs
impl Solution {
    pub fn check_inclusion(s1: String, s2: String) -> bool {
        let mut map1: [usize; 26] = [0; 26];
        let mut map2: [usize; 26] = [0; 26];
        for c in s1.chars() {
            map1[(c as u8 - b'a') as usize] += 1;
        }

        let n = s1.len();
        let s: &[u8] = s2.as_bytes();
        for (i, c) in s.iter().enumerate() {
            map2[(c - b'a') as usize] += 1;
            if i >= n {
                map2[(s[i - n] - b'a') as usize] -= 1;
            }

            if map1 == map2 {
                return true;
            }
        }

        false
    }
}
